### ISSUE TITLE
Enum UDT types not working from generator

### ISSUE THREAD

**Issue by tlf30** [impossibl/pgjdbc-ng#468]
Hello @kdubb 
I am working with some enums, and cannot get the UDT generator to build a working java object for either reading or writing. 

Here is my stacktrace:
```
**** Running Test3
Connecting to db...
Reading test data
com.impossibl.postgres.jdbc.PGSQLSimpleException: Error decoding column
	at com.impossibl.postgres.jdbc.PGResultSet.getObj(PGResultSet.java:304)
	at com.impossibl.postgres.jdbc.PGResultSet.getVal(PGResultSet.java:294)
	at com.impossibl.postgres.jdbc.PGResultSet.getObject(PGResultSet.java:891)
	at com.impossibl.postgres.jdbc.PGResultSet.getObject(PGResultSet.java:1092)
	at io.tlf.sqltest3.Main.test3(Main.java:65)
	at io.tlf.sqltest3.Main.main(Main.java:24)
Caused by: com.impossibl.postgres.system.ConversionException: Unable to convert value to class io.tlf.sqltest3.Condition
	at com.impossibl.postgres.system.procs.AutoConvertingBinaryDecoder.decodeValue(BaseDecoders.java:182)
	at com.impossibl.postgres.system.procs.BaseBinaryDecoder.decode(BaseDecoders.java:77)
	at com.impossibl.postgres.system.procs.Strings$BinDecoder.decode(Strings.java:179)
	at com.impossibl.postgres.system.procs.BaseBinaryDecoder.decode(BaseDecoders.java:47)
	at com.impossibl.postgres.protocol.v30.BufferRowData.getField(BufferRowData.java:157)
	at com.impossibl.postgres.jdbc.ListScroller.getRowField(PGResultSet.java:1858)
	at com.impossibl.postgres.jdbc.PGResultSet.getObj(PGResultSet.java:301)
	... 5 more
Writing test data
com.impossibl.postgres.jdbc.PGSQLSimpleException: invalid input value for enum condition: "GOOD"
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:197)
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:174)
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:161)
	at com.impossibl.postgres.jdbc.PGDirectConnection.execute(PGDirectConnection.java:509)
	at com.impossibl.postgres.jdbc.PGDirectConnection.executeTimed(PGDirectConnection.java:517)
	at com.impossibl.postgres.jdbc.PreparedQuery.executeStatement(PreparedQuery.java:119)
	at com.impossibl.postgres.jdbc.PreparedQuery.execute(PreparedQuery.java:186)
	at com.impossibl.postgres.jdbc.PGStatement.executeStatement(PGStatement.java:373)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.execute(PGPreparedStatement.java:339)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.executeLargeUpdate(PGPreparedStatement.java:373)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.executeUpdate(PGPreparedStatement.java:367)
	at io.tlf.sqltest3.Main.test3(Main.java:82)
	at io.tlf.sqltest3.Main.main(Main.java:24)

```

This is from my test case 3 which I uploaded here: https://github.com/tlf30/pgjdbc-ng-test3

The issue looks to be that perhaps the encoder and decoder are case insensitive, where postgres seems to be case sensitive on enum values. 

Any help would be very appreciated, I really love the UDT Generator and use it to create a LOT of object mappings!

Thank you, and Merry Christmas!
~Trevor

**Comment by tlf30** [comment]
@kdubb have you had a chance to look at this? I would really appreciate some help of getting this fixed. Currently I have had to stop using enums in my database

**Comment by kdubb** [comment]
Your referenced test appears to not be valid JDBC...
```java
  cond = result.getObject("cond", Condition.class);
```
 Specifically reading enums directly isn't supported by JDBC. You need to read a string and call `Condition.valueOf`.

If UDT generated code is doing something similar then its also incorrect. Do you have an example of that happening?

**Comment by tlf30** [comment]
Ah sorry, I got a bit to hasty when writing my test. (I was trying to catch a plane!).
I did mess up the read/write statements, that is what happens when I copy paste too fast :(

I have corrected the test to use an actual UDT generated read/write, which now makes a lot more sense as to what error I am seeing in my application. 
The UDT Generator is creating the following read (which works correctly):
```
@Override
  public void readSQL(SQLInput in, String typeName) throws SQLException {
    this.cond = Condition.valueOfLabel(in.readString());
  }
```

And the following write, which is what seems to have an issue:
```
  @Override
  public void writeSQL(SQLOutput out) throws SQLException {
    out.writeString(this.cond.getLabel());
  }
```

```
**** Running Test3
Connecting to db...
Reading test data
Writing test data
com.impossibl.postgres.jdbc.PGSQLSimpleException: wrong data type: 1043, expected 197978
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:197)
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:174)
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:161)
	at com.impossibl.postgres.jdbc.PGDirectConnection.execute(PGDirectConnection.java:509)
	at com.impossibl.postgres.jdbc.PGDirectConnection.executeTimed(PGDirectConnection.java:517)
	at com.impossibl.postgres.jdbc.PreparedQuery.executeStatement(PreparedQuery.java:119)
	at com.impossibl.postgres.jdbc.PreparedQuery.execute(PreparedQuery.java:186)
	at com.impossibl.postgres.jdbc.PGStatement.executeStatement(PGStatement.java:373)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.execute(PGPreparedStatement.java:339)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.executeLargeUpdate(PGPreparedStatement.java:373)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.executeUpdate(PGPreparedStatement.java:367)
	at io.tlf.sqltest3.Main.test3(Main.java:86)
	at io.tlf.sqltest3.Main.main(Main.java:24)
```

Where type 197978 is the CONDITION enum. 

Perhaps I am over looking something :D 

Anyways, sorry about the bad test, I believe it is correct now. 
Please let me know what you think. 

Thank you,
Trevor

**Comment by kdubb** [comment]
@tlf30 Can you try PR #470?  It writes the enum with its type specified I didn't have a chance to test it or add a test case yet.

**edit**
I know it generates compilable code as that is already covered in a test case but I haven't tried running it on a database with the correct schema.  If you are so inclined adding a PR for a test case in the `udt-gen` subproject that actually exercises the generated classes against the database would be a big help.

**Comment by tlf30** [comment]
I will build #470 and test it out. I'll try to write a test case today. I fly to work tomorrow and will be gone for two weeks. If I do not get one written today, I will write one when I get back.

**Comment by tlf30** [comment]
@kdubb I tried the PR, it took me a while to get working, for some reason this project never wants to publish to my ~/.m2 folder. 

This is the result of the run, looks the same to me:
```
**** Running Test3
Connecting to db...
Reading test data
Writing test data
com.impossibl.postgres.jdbc.PGSQLSimpleException: wrong data type: 1043, expected 197978
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:197)
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:174)
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:161)
	at com.impossibl.postgres.jdbc.PGDirectConnection.execute(PGDirectConnection.java:509)
	at com.impossibl.postgres.jdbc.PGDirectConnection.executeTimed(PGDirectConnection.java:517)
	at com.impossibl.postgres.jdbc.PreparedQuery.executeStatement(PreparedQuery.java:119)
	at com.impossibl.postgres.jdbc.PreparedQuery.execute(PreparedQuery.java:186)
	at com.impossibl.postgres.jdbc.PGStatement.executeStatement(PGStatement.java:373)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.execute(PGPreparedStatement.java:339)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.executeLargeUpdate(PGPreparedStatement.java:373)
	at com.impossibl.postgres.jdbc.PGPreparedStatement.executeUpdate(PGPreparedStatement.java:367)
	at io.tlf.sqltest3.Main.test3(Main.java:86)
	at io.tlf.sqltest3.Main.main(Main.java:24)
```

I have updated the test repo to include the binaries I built from the PR to be copied into your ~/.m2 to test with.

**Comment by kdubb** [comment]
Does you write method look like the below?

```
  @Override
  public void writeSQL(SQLOutput out) throws SQLException {
    out.writeObject(this.cond.getLabel(), Condition.TYPE);
  }
```

**Comment by tlf30** [comment]
No, this is what was generated:
```
  @Override
  public void readSQL(SQLInput in, String typeName) throws SQLException {
    this.cond = Condition.valueOfLabel(in.readString());
  }

  @Override
  public void writeSQL(SQLOutput out) throws SQLException {
    out.writeString(this.cond.getLabel());
  }
```

**Comment by kdubb** [comment]
It looks like you aren't using the code from the PR... it will generate it similar to what I posted and is doing so in the unit tests.

**Comment by tlf30** [comment]
Ah, even with changing the pom and running a clean install, IDEA was still caching the old dependencies on the classpath for running a single class. I will have to remember that one, I have never seen it do that before, I had to delete IDEA's project file and reimport. 

Anyways, the binaries I built are not working. It seems like they are not linking in the dependencies. Probably related to why they will not publish to maven local ~/.m2
